tests/cpp/test_learner.cc

```diff
@@ -115,4 +116,118 @@ TEST(Learner, SLOW_CheckMultiBatch) {
   learner->UpdateOneIter(0, dmat.get());
 }
 
+#if defined(XGBOOST_USE_CUDA)
+
+TEST(Learner, IO) {
+  using Arg = std::pair<std::string, std::string>;
+  size_t constexpr kRows = 10;
+  auto pp_dmat = CreateDMatrix(kRows, 10, 0);
+  auto p_dmat = *pp_dmat;
+
+  std::vector<bst_float> labels(kRows);
+  for (size_t i = 0; i < labels.size(); ++i) {
+    labels[i] = i;
+  }
+  p_dmat->Info().labels_.HostVector() = labels;
+  std::vector<std::shared_ptr<DMatrix>> mat {p_dmat};
+
+  std::unique_ptr<Learner> learner {Learner::Create(mat)};
+  learner->Configure({Arg{"predictor", "gpu_predictor"},
+                      Arg{"n_gpus", "-1"}});
+  learner->InitModel();
+  learner->UpdateOneIter(0, p_dmat.get());
+  ASSERT_EQ(learner->GetLearnerTrainParameter().gpu_id, 0);
+  ASSERT_EQ(learner->GetLearnerTrainParameter().n_gpus, -1);
+
+  dmlc::TemporaryDirectory tempdir;
+  const std::string fname = tempdir.path + "/model.bst";
+
+  {
+    // Create a scope to close the stream before next read.
+    std::unique_ptr<dmlc::Stream> fo(dmlc::Stream::Create(fname.c_str(), "w"));
+    learner->Save(fo.get());
+  }
+
+  std::unique_ptr<dmlc::Stream> fi(dmlc::Stream::Create(fname.c_str(), "r"));
+  learner->Load(fi.get());
+  ASSERT_EQ(learner->GetLearnerTrainParameter().gpu_id, 0);
+  ASSERT_EQ(learner->GetLearnerTrainParameter().n_gpus, 0);
```

Comment:
Why do we get 0 for `n_gpus` after loading the model? Is this because `tree_method` is `auto`?

So is it correct to say that you'll get CPU predictor when loading a learner with CPU algorithm, unless you explicitly override `predictor` parameter?

Reply:
@hcho3 Not that complicated.  Configuration is done after data is known, in `ConfugurationWithKnownData`.  Until the model see actual data, the default is CPU.  (Remember that this PR fixes XGBoost choosing GPU too aggressively).